README.md
# tap-quickbooks-report
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Python Versions](https://img.shields.io/badge/python-3.6%20%7C%203.7-blue.svg)](https://pypi.python.org/pypi/ansicolortags/)
[![Build Status](https://travis-ci.com/goodeggs/tap-quickbooks-report.svg?branch=master)](https://travis-ci.com/goodeggs/tap-quickbooks-report.svg?branch=master)

A [Singer](https://www.singer.io/) tap for extracting data from the Report Entities in the [Quickbooks Online API](https://developer.intuit.com/app/developer/qbo/docs/api/accounting/report-entities).

## Configuration

The Quickbooks Online API provides authentication via OAuth2.0. The tap expects the user to provide a valid Refresh Token via a config file. You can kick off
a OAuth2.0 User Consent process using the `--auth` argument while running the tap:

```bash
(tap-quickbooks-report) bash-3.2$ tap-quickbooks-report --auth
```

If you are an admin of the Quickbooks account, you will be able to authorize the Singer Tap application to access your account. You will then be redirected to a developer portal where you can access a valid Authentication Code and Realm ID. Copy and paste these values into the corresponding prompts in your terminal:

```bash
(tap-quickbooks-report) bash-3.2$ tap-quickbooks-report --auth
INFO Starting User Consent process..
Enter the Authorization Code: <auth-code>
Enter the Realm ID: <realm-id>
```

After entering the Authorization Code and Realm ID values, the tap will write the Realm ID, Refresh Token, and Refresh Token Expiration Date to the existing Config file:

```
(tap-quickbooks-report) bash-3.2$ tap-quickbooks-report --auth
INFO Starting User Consent process..
Enter the Authorization Code: <auth-code>
Enter the Realm ID: <realm-id>
INFO Generating new config..
```


CHANGELOG.md
## 0.1.2

 - Added fibonaccial decay backoff to the `_get()` method on the `QuickbooksStream`.
 - Made Rollbar logging more explicitly optional.
 - Loosened restrictions on required package versions.


tests/conftest.py
import argparse
import json

import pytest

from tap_quickbooks_report.streams import (ProfitAndLossDetailStream,
                                           ProfitAndLossStream)


@pytest.fixture(scope='function')
def config(shared_datadir):
    with open(shared_datadir / 'test.config.json') as f:
        return json.load(f)


@pytest.fixture(scope='function')
def args(config, shared_datadir):
    args = argparse.Namespace()
    setattr(args, 'config', config)
    setattr(args, 'config_path', shared_datadir / 'test.config.json')
    return args


@pytest.fixture(scope='function', params={ProfitAndLossStream, ProfitAndLossDetailStream})
def client(config, args, shared_datadir, request):
    return request.param(config=config,
                         args=args)


tests/test_client.py
import pytest
import requests
from singer.schema import Schema

from tap_quickbooks_report.streams import is_fatal_code


@pytest.mark.parametrize('status_code', [400, 401, 403, 404,
                                         pytest.param(500, marks=pytest.mark.xfail),
                                         pytest.param(502, marks=pytest.mark.xfail),
                                         pytest.param(503, marks=pytest.mark.xfail),
                                         pytest.param(504, marks=pytest.mark.xfail)])
def test_is_fatal_code(status_code):
    resp = requests.models.Response()
    resp.status_code = status_code
    exc = requests.exceptions.RequestException(response=resp)
    assert is_fatal_code(exc)


def test_load_schema(client):
    schema = client._load_schema()
    assert isinstance(schema, dict)
    assert Schema.from_dict(schema)


tap_quickbooks_report/utils.py
import argparse
import json


def load_json(path):
    with open(path) as fil:
        return json.load(fil)


def parse_args(required_config_keys):
    '''Parse standard command-line args.
    Parses the command-line arguments mentioned in the SPEC and the
    BEST_PRACTICES documents:
    -c,--config     Config file
    -s,--state      State file
    -d,--discover   Run in discover mode
    -p,--properties Properties file: DEPRECATED, please use --catalog instead
    -a,--auth       Establish user consent to retrieve OAuth2.0 credentials
    --catalog       Catalog file
    Returns the parsed args object from argparse. For each argument that
    point to JSON files (config, state, properties), we will automatically
    load and parse the JSON file.
    '''
    parser = argparse.ArgumentParser()

    parser.add_argument(
        '-c', '--config',
        help='Config file',
        required=True)

    parser.add_argument(
        '-s', '--state',
        help='State file')

    parser.add_argument(
        '-p', '--properties',
        help='Property selections: DEPRECATED, Please use --catalog instead')

    parser.add_argument(
        '--catalog',
        help='Catalog file')

    parser.add_argument(
        '-d', '--discover',
        action='store_true',
        help='Do schema discovery')

    parser.add_argument(
        '-a', '--auth',
        action='store_true',
        help='Establish user consent to retrieve OAuth2.0 credentials')

    args = parser.parse_args()
    if args.config:
        setattr(args, 'config_path', args.config)
        args.config = load_json(args.config)
    if args.state:
        setattr(args, 'state_path', args.state)
        args.state = load_json(args.state)
    else:
        args.state = {}
    if args.properties:
        setattr(args, 'properties_path', args.properties)
        args.properties = load_json(args.properties)

    check_config(args.config, required_config_keys)

    return args


def check_config(config, required_keys):
    missing_keys = [key for key in required_keys if key not in config]
    if missing_keys:
        raise Exception("Config is missing required keys: {}".format(missing_keys))


tap_quickbooks_report/__init__.py
import os

import rollbar
import singer

from .streams import (ProfitAndLossDetailStream, ProfitAndLossStream,
                      QuickbooksStream)
from .utils import parse_args

LOGGER = singer.get_logger()


try:
    ROLLBAR_ACCESS_TOKEN = os.environ["ROLLBAR_ACCESS_TOKEN"]
    ROLLBAR_ENVIRONMENT = os.environ["ROLLBAR_ENVIRONMENT"]
except KeyError:
    LOGGER.info("No Rollbar environment variables found. Rollbar logging disabled..")
    log_to_rollbar = False
else:
    rollbar.init(ROLLBAR_ACCESS_TOKEN, ROLLBAR_ENVIRONMENT)
    log_to_rollbar = True

AUTH_REQUIRED_CONFIG_KEYS = [
    "client_id",
    "client_secret",
    "environment",
    "redirect_uri"
]

SYNC_REQUIRED_CONFIG_KEYS = [
    "client_id",
    "client_secret",
    "environment",
    "redirect_uri",
    "realm_id",
    "refresh_token",
    "refresh_token_expires_at"
]


def user_consent(config, args):
    LOGGER.info('Starting User Consent process..')
    stream = QuickbooksStream(config=config, args=args)
    stream.user_consent()


def sync(config, args):
    LOGGER.info('Starting sync..')
    stream = ProfitAndLossStream(config=config, args=args)
    stream.write_schema_message()
    stream.sync()
    stream = ProfitAndLossDetailStream(config=config, args=args)
    stream.write_schema_message()
    stream.sync()


def main():
    args = parse_args(required_config_keys=AUTH_REQUIRED_CONFIG_KEYS)
    if args.auth is True:
        try:
            user_consent(config=args.config, args=args)
        except:
            LOGGER.exception('Caught exception during User Consent..')
            if log_to_rollbar is True:
                rollbar.report_exc_info()
    else:
        args = parse_args(required_config_keys=SYNC_REQUIRED_CONFIG_KEYS)
        try:
            sync(config=args.config, args=args)
        except:
            LOGGER.exception('Caught exception during Sync..')
            if log_to_rollbar is True:
                rollbar.report_exc_info()


if __name__ == "__main__":
    main()


tap_quickbooks_report/streams.py
import json
import os
import webbrowser
from datetime import datetime, timedelta
from typing import ClassVar, Dict, List, Optional

import attr
import backoff
import pytz
import requests
import singer
from intuitlib.client import AuthClient
from intuitlib.enums import Scopes

from .version import __version__

LOGGER = singer.get_logger()


def is_fatal_code(e: requests.exceptions.RequestException) -> bool:
    '''Helper function to determine if a Requests reponse status code
    is a "fatal" status code. If it is, the backoff decorator will giveup
    instead of attemtping to backoff.'''
    return 400 <= e.response.status_code < 500 and e.response.status_code != 429


@attr.s
class QuickbooksStream:
    tap_stream_id: ClassVar[Optional[str]] = None
    base_url: ClassVar[str] = "https://quickbooks.api.intuit.com"
    api_version: ClassVar[str] = "v3"
    api_minor_version: ClassVar[int] = 40

    config: Dict = attr.ib()
    args: Dict = attr.ib()
    @config.validator
    def check(self, attribute, value):
        if value.get("environment") not in ["sandbox", "production"]:
            raise ValueError('environment attribute must be either "sandbox" or "production".')

    def _get_abs_path(self, path: str) -> str:
        return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)

    def _load_schema(self) -> Dict:
        '''Loads a JSON schema file for a given
        Quickbooks Report resource into a dict representation.
        '''
        schema_path = self._get_abs_path("schemas")
        return singer.utils.load_json(f"{schema_path}/{self.tap_stream_id}.json")

    def _generate_token_expiration(self, refresh_token_expires_in_seconds: int) -> str:
        '''Generates a string-formatted expiration date using the
        refresh_token_expires_in_seconds value attached to the Auth Client.
        '''
        return datetime.strftime((datetime.utcnow().replace(tzinfo=pytz.UTC) + timedelta(seconds=refresh_token_expires_in_seconds)), '%Y-%m-%d %H:%M:%S %Z')

    def user_consent(self):
        '''Triggers the User Consent OAuth2.0 flow
        in order to retrieve an Authorization Code
        and a Realm ID.
        '''
        auth_client = AuthClient(self.config.get("client_id"),
                                 self.config.get("client_secret"),
                                 self.config.get("redirect_uri"),
                                 self.config.get("environment"))

        scopes = [Scopes.ACCOUNTING]
        auth_url = auth_client.get_authorization_url(scopes)
        webbrowser.open(auth_url)
        auth_code = input('Enter the Authorization Code: ')
        realm_id = input('Enter the Realm ID: ')
        auth_client.get_bearer_token(auth_code, realm_id=realm_id)
        self.config["realm_id"] = realm_id
        self.config["refresh_token"] = auth_client.refresh_token
        self.config["refresh_token_expires_at"] = self._generate_token_expiration(auth_client.x_refresh_token_expires_in)

        LOGGER.info('Generating new config..')
        with open(self.args.config_path, 'r+') as f:
            json.dump(self.config, f, indent=2)

    def _check_token_expiry(self, auth_client):
        '''Checks the expiration status
        of the Refresh Token supplied in the config file.
        '''
        LOGGER.info("Checking Refresh Token expiry..")
        rt_expires_in = timedelta(seconds=auth_client.x_refresh_token_expires_in)
        rt_expiration_dt_utc = datetime.utcnow() + rt_expires_in
        refresh_token_msg = f"Refresh Token expires on {rt_expiration_dt_utc} UTC"
        if rt_expires_in.days <= 30:
            LOGGER.warning(refresh_token_msg)
        else:
            LOGGER.info(refresh_token_msg)

    def _get_auth_client(self):
        '''Returns an OAuth2.0 Client for interacting
        with Quickbooks Reporting API.
        '''

        auth_client = AuthClient(self.config.get("client_id"),
                                 self.config.get("client_secret"),
                                 self.config.get("redirect_uri"),
                                 self.config.get("environment"),
                                 refresh_token=self.config.get("refresh_token"),
                                 realm_id=self.config.get("realm_id"))

        # Refresh to get new Access Token.
        auth_client.refresh()

        if auth_client.refresh_token == self.config.get("refresh_token"):
            LOGGER.info("Config file Refresh Token and Refresh Token received from Refresh Token API are identical.")
        else:
            LOGGER.info("Config file Refresh Token and Refresh Token received from Refresh Token API has drifted.")
            LOGGER.info("Overwriting Config file with new Refresh Token values..")

            self.config["refresh_token"] = auth_client.refresh_token
            self.config["refresh_token_expires_at"] = self._generate_token_expiration(auth_client.x_refresh_token_expires_in)

            with open(self.args.config_path, 'r+') as f:
                json.dump(self.config, f, indent=2)

        # Check Refresh Token Expiry.
        self._check_token_expiry(auth_client)

        return auth_client

    def _construct_headers(self, access_token) -> Dict:
        '''Constructs a standard set of headers for GET requests.'''
        headers = requests.utils.default_headers()
        headers["Accept"] = "application/json"
        headers["User-Agent"] = f"python-quickbooks-reporting-tap/{__version__}"
        headers["Authorization"] = f"Bearer {access_token}"
        headers["Content-Type"] = "application/json"
        headers["Date"] = singer.utils.strftime(singer.utils.now(), '%a, %d %b %Y %H:%M:%S %Z')
        return headers

    @backoff.on_exception(backoff.fibo,
                          requests.exceptions.HTTPError,
                          max_time=120,
                          giveup=is_fatal_code,
                          logger=LOGGER)
    @backoff.on_exception(backoff.fibo,
                          (requests.exceptions.ConnectionError,
                           requests.exceptions.Timeout),
                          max_time=120,
                          logger=LOGGER)
    def _get(self, auth_client, report_entity: str, params: Optional[Dict] = None) -> Dict:
        '''Constructs a standard way of making
        a GET request to the Quickbooks REST API.
        '''
        url = f"{self.base_url}/{self.api_version}/company/{auth_client.realm_id}/reports/{report_entity}"
        headers = self._construct_headers(access_token=auth_client.access_token)
        if params:
            params.update({"minorversion": self.api_minor_version})
        response = requests.get(url, headers=headers, params=params)
        response.raise_for_status()
        return response.json()

    def _convert_string_value_to_float(self, value: str) -> float:
        '''Safely converts string values to floats.'''
        if value == "":
            float_value = float(0.0)
        else:
            float_value = float(value)
        return float_value

    def _get_row_data(self, resp, column_enum, input):
        for row in resp.get("Rows").get("Row"):
            if row.get("type") == "Section" and row.get("Rows") is not None:
                data_dict = {}
                key = row.get("Header").get("ColData")[0].get("value").title().replace(" ", "")
                data = {
                    key: {
                        "Lines": [],
                        "Total": self._convert_string_value_to_float(row.get("Summary").get("ColData")[column_enum].get("value"))
                    }
                }
                data_dict.update(data)
                input.append(data_dict)
                self._get_row_data(resp=row, input=data_dict[key]["Lines"], column_enum=column_enum)
            elif row.get("Summary") is not None:
                data_dict = {}
                key = row.get("Summary").get("ColData")[0].get("value").title().replace(" ", "")
                data = {
                    key: {
                        "Lines": [],
                        "Total": self._convert_string_value_to_float(row.get("Summary").get("ColData")[column_enum].get("value"))
                    }
                }
                data_dict.update(data)
                input.append(data_dict)
            else:
                data_dict = {}
                key = row.get("ColData")[0].get("value").title().replace(" ", "")
                data = {
                    key: {
                        "Lines": [],
                        "Total": self._convert_string_value_to_float(row.get("ColData")[column_enum].get("value"))
                    }
                }
                data_dict.update(data)
                input.append(data_dict)

        return input

    def _transform_columns_into_rows(self, resp):
        records = []
        for column in resp.get("Columns").get("Column"):
            if column.get("ColType") == "Money":
                record = {}
                for meta_column in column.get("MetaData"):
                    record[meta_column.get("Name")] = meta_column.get("Value")
                records.append(record)

        return records

    def write_schema_message(self):
        '''Writes a Singer schema message.'''
        return singer.write_schema(stream_name=self.stream, schema=self.schema, key_properties=self.key_properties)


class ProfitAndLossStream(QuickbooksStream):
    tap_stream_id: ClassVar[str] = 'profit_and_loss'
    stream: ClassVar[str] = 'profit_and_loss'
    key_properties: ClassVar[str] = 'StartDate'
    replication_method: ClassVar[str] = 'FULL_TABLE'

    def __init__(self, config: Dict, args: Dict):
        self.schema = self._load_schema()
        super().__init__(config, args)

    def sync(self):
        with singer.metrics.job_timer(job_type=f"sync_{self.tap_stream_id}"):
            with singer.metrics.record_counter(endpoint=self.tap_stream_id) as counter:
                client = self._get_auth_client()
                params = {
                    "start_date": "2014-01-01",
                    "accounting_method": "Accrual",
                    "summarize_column_by": "Month"
                }
                resp = self._get(auth_client=client, report_entity='ProfitAndLoss', params=params)
                rows = self._transform_columns_into_rows(resp)

                for i, row in enumerate(rows):
                    if row.get("StartDate") is None:
                        continue
                    input = []
                    data = self._get_row_data(resp=resp, column_enum=i + 1, input=input)
                    new_data = {}
                    for line in data:
                        new_data.update(line)
                    row["ReportData"] = new_data
                    row["SyncTimestampUtc"] = singer.utils.strftime(singer.utils.now(), "%Y-%m-%dT%H:%M:%SZ")

                    with singer.Transformer() as transformer:
                        transformed_record = transformer.transform(data=row, schema=self.schema)
                        singer.write_record(stream_name=self.stream, time_extracted=singer.utils.now(), record=transformed_record)
                        counter.increment()


class ProfitAndLossDetailStream(QuickbooksStream):
    tap_stream_id: ClassVar[str] = 'profit_and_loss_detail'
    stream: ClassVar[str] = 'profit_and_loss_detail'
    key_properties: ClassVar[List[str]] = []
    replication_method: ClassVar[str] = 'FULL_TABLE'

    def __init__(self, config: Dict, args: Dict):
        self.schema = self._load_schema()
        super().__init__(config, args)

    def _get_column_metadata(self, resp):
        columns = []
        for column in resp.get("Columns").get("Column"):
            if column.get("ColTitle") == "Memo/Description":
                columns.append("Memo")
            else:
                columns.append(column.get("ColTitle").replace(" ", ""))
        columns.append("Categories")
        return columns

    def _recursive_row_search(self, row, output, categories):
        row_group = row.get("Rows")
        if 'ColData' in list(row.keys()):
            # Write the row
            data = row.get("ColData")
            values = [column.get("value") for column in data]
            categories_copy = categories.copy()
            values.append(categories_copy)
            values_copy = values.copy()
            output.append(values_copy)
        elif row_group is None or row_group == {}:
            pass
        else:
            row_array = row_group.get("Row")
            header = row.get("Header")
            if header is not None:
                categories.append(header.get("ColData")[0].get("value"))
            for row in row_array:
                self._recursive_row_search(row, output, categories)
            if header is not None:
                categories.pop()

    def sync(self):
        singer_version = int(datetime.utcnow().timestamp())
        with singer.metrics.job_timer(job_type=f"sync_{self.tap_stream_id}"):
            with singer.metrics.record_counter(endpoint=self.tap_stream_id) as counter:
                client = self._get_auth_client()
                params = {
                    "start_date": "2014-01-01",
                    "accounting_method": "Accrual"
                }
                resp = self._get(auth_client=client, report_entity='ProfitAndLossDetail', params=params)

                # Get column metadata.
                columns = self._get_column_metadata(resp)

                # Recursively get row data.
                row_group = resp.get("Rows")
                row_array = row_group.get("Row")

                if row_array is None:
                    return LOGGER.info("Report has no rows!")

                output = []
                categories = []
                for row in row_array:
                    self._recursive_row_search(row, output, categories)

                # Zip columns and row data.
                for raw_row in output:
                    row = dict(zip(columns, raw_row))
                    cleansed_row = {}
                    for k, v in row.items():
                        if v == "":
                            continue
                        else:
                            cleansed_row.update({k: v})

                    cleansed_row["Amount"] = float(row.get("Amount"))
                    cleansed_row["Balance"] = float(row.get("Balance"))
                    cleansed_row["SyncTimestampUtc"] = singer.utils.strftime(singer.utils.now(), "%Y-%m-%dT%H:%M:%SZ")

                    with singer.Transformer() as transformer:
                        transformed_record = transformer.transform(data=cleansed_row, schema=self.schema)
                        singer.write_message(singer.RecordMessage(stream=self.stream,
                                                                  record=transformed_record,
                                                                  version=singer_version,
                                                                  time_extracted=singer.utils.now()))
                        counter.increment()

        singer.write_version(stream_name=self.stream, version=singer_version)
